bolt-design-system/bolt
# Add Bolt Docs Site Search (Powered By Algolia)

## Jira Ticket
http://vjira2:8080/browse/BDS-593

## Summary
Adds search to the Bolt documentation site, powered by Algolia.

![F160EA75-A490-447D-AAC6-7059C12CECC1](https://user-images.githubusercontent.com/1617209/55151786-62847b00-5125-11e9-8fa8-a5fbec506d9d.gif)

## Details
- Packages up the default Algolia docs search / autocomplete library in a tiny new `<bolt-docs-search>` package so we don’t need to load everything via a different domain / different CDN
- Adds the ability to CMS + click search results to open in a new tab
- In lieu of not having our current form search input be available as a web component, this docs site specific component instead async renders via the component’s Twig template + our Twig rendering service
- Adds a missing rendering service API endpoint to allow us to render Twig template strings + temples using Twig extensions
- Adds temporary docs site specific style overrides to the default Bolt search input / icon styles used, given that the current form input components (seemingly) provides no way to:
    - Include a smaller / more compact version of our search inputs
    - Disable requiring an error icon (which takes up space)
    - Disable including an icon background color
    - Change the position of the search icon (to display on the left vs right side)
    - Disable the gradient fade between the input and the icon
    - Round the corners to the search input
    - Allow the search input to automatically expand / collapse in width depending on if the nested input within has `:focus`
    - Enable the default browser clear search button
    - Disable the search icon from always being a submit button (dynamic AJAX search like this one can’t use a traditional submit button)
- In addition to looking into incorporating these config and style enhancements to our current form components, this PR also adds a new optional `icon_size` config option to the form component to provide a way to optionally configure the icon of the icons automatically included in our form components.

> Note: for reference, the [docs site search config](https://github.com/algolia/docsearch-configs/blob/master/configs/boltdesignsystem.json) lives on Algolia’s config repo if / when we need to fine tune the URLs that get indexed + the selectors getting used. The current config points to the main boltdesignsystem.com site’s `/docs` section by default which, if we wanted full component docs to get listed, would require this config to get updated, our site structure to get updated, or our docs to get updated.

## How to test
- Confirm search working as expected
- Confirm existing form component UI (specifically components with icons) are rendering as expected with no visible changes
